### PR TITLE
[Bug] Fail on non-static range in backwards

### PR DESCRIPTION
**Pull request by hughperkins**
Issue: #

### Brief Summary

copilot:summary

### Walkthrough

copilot:walkthrough

### CODE REVIEW

**Comment by SonSang** [comment]
AD does not work properly in the backward pass when non-static `inner` loops are used. However, the results are quite inconsistent and unpredictable, so I summarize here some cases that I have faced so far.

* Compiles and runs silently, but memory usage is largely increased
* Compiles and runs silently, but gives `nan` value
* Compiles and runs silently, but gradient is not computed correctly
* Does not compile at all

Here is the example where it does not compile at all:
```
import gstaichi as ti

ti.init(arch=ti.cpu, debug=True)

B = 10
x = ti.field(ti.f32, shape=(3, B), needs_grad=True)
y = ti.field(ti.f32, shape=(3, B), needs_grad=True)
loss = ti.field(ti.f32, shape=(), needs_grad=True)

x[0, 0] = 1.0
y[0, 0] = 2.0

@ti.kernel
def k1(use_static: ti.template()):
    for i_b in ti.ndrange(B):
        z = ti.Vector.zero(ti.f32, 3)
        
        # Non-static inner loop is not supported in backward
        for i_3 in ti.static(range(3)) if ti.static(use_static) else range(3):
            z[i_3] += x[i_3, i_b] * y[i_3, i_b]

        loss[None] += z.x + z.y + z.z

for i in range(2):
    try:
        use_static = True if i == 0 else False

        loss.fill(0.0)
        k1(use_static) 
        
        loss.grad[None] = 1.0
        x.grad.fill(0.0)
        y.grad.fill(0.0)
        k1.grad(use_static)

        assert x.grad[0,0] == 2.0
        assert y.grad[0,0] == 1.0

        print("Passed for use_static =", use_static)
        input("Press Enter to continue...")
    except Exception as e:
        print("Failed for use_static =", use_static)
        if i == 1:
            print("This is expected failure, because non-static inner loop is not supported")
            print("Error:", e)
```

**Comment by hughperkins** [comment]
> AD does not work properly in the backward pass when non-static `inner` loops are used. However, the results are quite inconsistent and unpredictable, so I summarize here some cases that I have faced so far.
> 
> * Compiles and runs silently, but memory usage is largely increased
> * Compiles and runs silently, but gives `nan` value
> * Compiles and runs silently, but gradient is not computed correctly
> * Does not compile at all
> 
> Here is the example where it does not compile at all:
> 
> ```
> import gstaichi as ti
> 
> ti.init(arch=ti.cpu, debug=True)
> 
> B = 10
> x = ti.field(ti.f32, shape=(3, B), needs_grad=True)
> y = ti.field(ti.f32, shape=(3, B), needs_grad=True)
> loss = ti.field(ti.f32, shape=(), needs_grad=True)
> 
> x[0, 0] = 1.0
> y[0, 0] = 2.0
> 
> @ti.kernel
> def k1(use_static: ti.template()):
>     for i_b in ti.ndrange(B):
>         z = ti.Vector.zero(ti.f32, 3)
>         
>         # Non-static inner loop is not supported in backward
>         for i_3 in ti.static(range(3)) if ti.static(use_static) else range(3):
>             z[i_3] += x[i_3, i_b] * y[i_3, i_b]
> 
>         loss[None] += z.x + z.y + z.z
> 
> for i in range(2):
>     try:
>         use_static = True if i == 0 else False
> 
>         loss.fill(0.0)
>         k1(use_static) 
>         
>         loss.grad[None] = 1.0
>         x.grad.fill(0.0)
>         y.grad.fill(0.0)
>         k1.grad(use_static)
> 
>         assert x.grad[0,0] == 2.0
>         assert y.grad[0,0] == 1.0
> 
>         print("Passed for use_static =", use_static)
>         input("Press Enter to continue...")
>     except Exception as e:
>         print("Failed for use_static =", use_static)
>         if i == 1:
>             print("This is expected failure, because non-static inner loop is not supported")
>             print("Error:", e)
> ```

This PR is aiming to address the examples where it does compile. Since if it does not compile, then behavior is correct already, and does not need to be modified.

Could you provide an example please for each of:
* Compiles and runs silently, but memory usage is largely increased
* Compiles and runs silently, but gives `nan` value
* Compiles and runs silently, but gradient is not computed correctly

(or at least one of these, ideally)

File: python/gstaichi/lang/ast/ast_transformer_utils.py [hunk 1]
```diff
@@ -199,6 +204,17 @@ def __init__(
         self.is_real_function = is_real_function
         self.kernel_args: list = []
         self.only_parse_function_def: bool = False
+        self.autodiff_mode = autodiff_mode
+        self.loop_depth: int = 0
+        from gstaichi import extension  # pylint: disable=import-outside-toplevel
```
Comment (duburcqa) [inline python/gstaichi/lang/ast/ast_transformer_utils.py]: I prefer keeping local import on top of the corresponding scope / code block (in this case, on top of the function definition). I find it much more readable. But do as you please if you don't share my opinion.

Comment (hughperkins) [reply python/gstaichi/lang/ast/ast_transformer_utils.py]: It broke the linter, since it said that torch might not be bound.

Open to alternative approaches to fix the linter.

Comment (duburcqa) [reply python/gstaichi/lang/ast/ast_transformer_utils.py]: I don't understand. I'm only suggesting a minor change that should not have any effect on the linter:
```python
        is_real_function: bool,
        autodiff_mode: AutodiffMode,
    ):
        from gstaichi import extension

        self.func = func
        self.local_scopes: list[dict[str, Any]] = []
        [...]
```

File: tests/python/test_ad_for.py [hunk 2]
```diff
@@ -3,10 +3,13 @@
 from tests import test_utils
 
 
-@test_utils.test()
+@test_utils.test(require=ti.extension.adstack)
 def test_ad_nested_for():
     N = 5
 
+    assert ti.lang is not None
+    ti.lang.impl.current_cfg().ad_stack_experimental_enabled = True
```
Comment (duburcqa) [inline tests/python/test_ad_for.py]: `@test_utils.test` should do `ti.lang.impl.current_cfg().ad_stack_experimental_enabled = True` automatically whenever `ti.extension.adstack` is required.

The point of pytest vs unittest is just to expose facilities to hide all the boilerplate, for devs and readers to focus on what really matters. As a rule of thumb, if you need to copy-paste the same code in multiple test, you probably need a fixture to do this for you behind the scene without you paying attention to it.

File: tests/python/test_ad_ndarray.py [hunk 3]
```diff
@@ -32,6 +31,8 @@ def test(x: ti.types.ndarray(), y: ti.types.ndarray()):
 @pytest.mark.skipif(not has_pytorch(), reason="Pytorch not installed.")
 @test_utils.test(arch=archs_support_ndarray_ad, default_fp=ti.f64)
 def test_ad_reduce():
+    import torch
```
Comment (duburcqa) [inline tests/python/test_ad_ndarray.py]: What is this?

Comment (hughperkins) [reply tests/python/test_ad_ndarray.py]: otherwise there are linter errors about 'torch might be unbound'

File: tests/python/test_ad_ndarray.py [hunk 4]
```diff
@@ -6,15 +6,14 @@
 
 from tests import test_utils
 
-if has_pytorch():
-    import torch
-
```
Comment (duburcqa) [inline tests/python/test_ad_ndarray.py:11]: What about moving all the tests that requires torch in a dedicated test file and do:
```python
try:
    import torch
except ImportError:
    pytest.skip("PyTorch not installed. Skipping...", allow_module_level=True)
```

Comment (hughperkins) [reply tests/python/test_ad_ndarray.py:11]: addressed in 546998697
